packages/sources/gsr/src/transport/authutils.ts

```diff
@@ -0,0 +1,52 @@
+import crypto from 'crypto'
+import axios from 'axios'
+import { config } from '../config'
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+
+const logger = makeLogger('GSR Auth Token Utils')
+
+interface TokenError {
+  success: false
+  ts: number
+  error: string
+}
+
+interface TokenSuccess {
+  success: true
+  ts: number
+  token: string
+  validUntil: string
+}
+
+type AccessTokenResponse = TokenError | TokenSuccess
+
+const currentTimeNanoSeconds = (): number => new Date(Date.now()).getTime() * 1000000
+
+const generateSignature = (userId: string, publicKey: string, privateKey: string, ts: number) =>
+  crypto
+    .createHmac('sha256', privateKey)
+    .update(`userId=${userId}&apiKey=${publicKey}&ts=${ts}`)
+    .digest('hex')
+
+export const getToken = async (settings: typeof config.settings) => {
+  logger.debug('Fetching new access token')
+
+  const userId = settings.WS_USER_ID
+  const publicKey = settings.WS_PUBLIC_KEY
+  const privateKey = settings.WS_PRIVATE_KEY
+  const ts = currentTimeNanoSeconds()
+  const signature = generateSignature(userId, publicKey, privateKey, ts)
+  const response = await axios.post<AccessTokenResponse>(`${settings.API_ENDPOINT}/token`, {
+    apiKey: publicKey,
+    userId,
+    ts,
+    signature,
+  })
+
+  if (!response.data.success) {
+    logger.error('Unable to get access token')
+    throw new Error(response.data.error)
```

Comment:
Does this error contain a convenient string we could include in our `logger.error`? Would make future debugging slightly easier

Appreciate you've just moved this rather than changed, so feel free to leave as is

Reply:
are you suggesting to add the `error` field into the log?

a few example failure cases:
`{"ts":1692108328384919770,"success":false,"error":"UserID not found","token":null,"validUntil":null}`
`{"ts":1692108303988933652,"success":false,"error":"API key mismatch"}`
`{"ts":1692044967180622754,"success":false,"error":"Signature is required"}`

Reply:
reread your original comment, will add it, though I'd assume thrown errors are logged at some point as well